src/main/java/org/phyloref/jphyloref/commands/WebserverCommand.java

```diff
@@ -0,0 +1,453 @@
+import org.semanticweb.owlapi.model.OWLClass;
+import org.semanticweb.owlapi.model.OWLNamedIndividual;
+import org.semanticweb.owlapi.model.OWLOntology;
+import org.semanticweb.owlapi.model.OWLOntologyCreationException;
+import org.semanticweb.owlapi.model.OWLOntologyLoaderConfiguration;
+import org.semanticweb.owlapi.model.OWLOntologyManager;
+import org.semanticweb.owlapi.reasoner.BufferingMode;
+import org.semanticweb.owlapi.rio.RioOWLRDFConsumerAdapter;
+import org.semanticweb.owlapi.search.EntitySearcher;
+import org.semanticweb.owlapi.util.AnonymousNodeChecker;
+import org.semanticweb.owlapi.util.VersionInfo;
+
+import fi.iki.elonen.NanoHTTPD;
+import fi.iki.elonen.NanoHTTPD.Response.Status;
+import fi.iki.elonen.NanoHTTPD.Response.IStatus;
+import uk.ac.manchester.cs.jfact.JFactReasoner;
+import uk.ac.manchester.cs.jfact.kernel.options.JFactReasonerConfiguration;
+
+/**
+ * Sets up a webserver that allows reasoning over phyloreferences
+ * over HTTP.
+ *
+ * At the moment, we implement a simple API, in which:
+ *  /version:   returns a JSON object with information on the version of
+ *              JPhyloRef and the reasoner being used.
+ *  /reason:    expects a form upload with a 'jsonld' element containing a file
+ *              upload of a JSON-LD file representing an ontology to test.
+ *              This command will reason over the ontology and return a JSON
+ *              dictionary, in which the keys are the IRIs for each phyloreference,
+ *              and the values are lists of the IRIs of each node matched by that
+ *              phyloreference.
+ *
+ * @author Gaurav Vaidya <gaurav@ggvaidya.com>
+ *
+ */
+public class WebserverCommand implements Command {
+  /**
+   * This command is named "webserver". It should be
+   * invoked as "java -jar jphyloref.jar webserver ..."
+   */
+  @Override
+  public String getName() {
+    return "webserver";
+  }
+
+  /**
+   * A description of the Webserver command.
+   *
+   * @return A description of this command.
+   */
+  @Override
+  public String getDescription() {
+    return "Set up a webserver to allow reasoning of phyloreferences over HTTP.";
+  }
+
+  /**
+   * Add command-line options specific to this command.
+   *
+   * @param opts The command-line options to modify for this command.
+   */
+  @Override
+  public void addCommandLineOptions(Options opts) {
+    opts.addOption(
+      "h", "host", true,
+      "The hostname to listen to HTTP connections on (default: 'localhost')"
+    );
+    opts.addOption(
+      "p", "port", true,
+      "The TCP port to listen to HTTP connections on (default: 8080)"
+    );
+  }
+
+  /**
+   * Set up a webserver to listen on the provided hostname and port (or their defaults).
+   *
+   * @param cmdLine The command line options provided to this command.
+   */
+  @Override
+  public void execute(CommandLine cmdLine) throws RuntimeException {
+      String hostname = cmdLine.getOptionValue("host", "localhost");
```

Comment:
I set it up this way because I remember some issues we ran into with Jekyll where it would run on "127.0.0.1" on Windows and on "localhost" on Linux; this way, the user can choose which of the available IP addresses or hostnames they want to use. I've just confirmed that this works on locally usable hostnames and addresses (e.g. localhost, 127.0.0.3) but not on remote hostnames or IP addresses (e.g. phyloref.org, 8.8.8.8). Is there a security loophole I'm not seeing here?

Reply:
127.0.0.1 should work universally. (In Unix, and thus Linux, the network interface (the loopback device) associated with  `localhost` and `127.0.0.1` is exactly the same.)

I guess you could argue that there's valid possibility of wanting to run it in a different container on the same virtual network that allows a set of containers (but nobody else) to talk to each other.

Reply:
I looked around to see if there were any other differences between the two, and I did come up with one: `localhost` can map to `[::1]` on IPv6, in which case the socket won't listen on `127.0.0.1` at all. I don't think we need to design for supporting both IPv4 and IPv6, but I like that having an `-h` command line option allows the user to pick which one they want!